File: Week 7/cpp/1.string_basics.cpp
/*
 * WEEK 7 - C++ DSA
 * Topic: String Basics
 * File: 1.string_basics.cpp
 *
 * CONCEPT:
 *     std::string is a mutable, contiguous container of char.
 *     Common operations: size, indexing, substr, find, replace,
 *     stream concatenation with std::ostringstream / std::string +=.
 *
 * KEY POINTS:
 *     - std::string is mutable, unlike Java String — append in place is O(1)
 *       amortised.
 *     - std::string::find returns std::string::npos when not found.
 *     - For repeated building, std::ostringstream / push_back / += all work;
 *       += is O(1) amortised because std::string maintains a capacity buffer.
 *
 * ALGORITHM / APPROACH:
 *     Showcase: size(), [i], substr(start, length), find(c), find(s),
 *     replace, transform (std::transform + std::tolower), split.
 *
 * C++-SPECIFIC NOTES:
 *     - No built-in split; show a manual one using std::stringstream.
 *     - std::transform with ::tolower converts a string in place.
 *     - To reverse, use std::reverse on the iterator pair.
 *
 * DRY RUN:
 *     s = "Hello, World!"
 *     s.size() = 13
 *     s[7] = 'W'
 *     s.substr(7, 5) = "World"
 *     s.find('o') = 4
 *
 * COMPLEXITY:
 *     Most operations O(n) in the size of their argument; subscript is O(1).
 */

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>
#include <cctype>

std::vector<std::string> split(const std::string& s, char delim) {
    std::vector<std::string> out;
    std::string token;
    std::stringstream ss(s);
    while (std::getline(ss, token, delim)) out.push_back(token);
    return out;
}

std::string toLower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return std::tolower(c); });
    return s;
}

std::string toUpper(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return std::toupper(c); });
    return s;
}

int main() {
    std::string s = "Hello, World!";
    std::cout << "String: " << s << "\n";
    std::cout << "size: " << s.size() << "\n";
    std::cout << "s[7]: " << s[7] << "\n";
    std::cout << "substr(7, 5): " << s.substr(7, 5) << "\n";
    std::cout << "find('o'): " << s.find('o') << "\n";
    std::cout << "rfind('o'): " << s.rfind('o') << "\n";
    std::cout << "find(\"World\"): " << s.find("World") << "\n";
    std::cout << "toLower: " << toLower(s) << "\n";
    std::cout << "toUpper: " << toUpper(s) << "\n";

    // Replace all 'l' with 'r'
    std::string r = s;
    std::replace(r.begin(), r.end(), 'l', 'r');
    std::cout << "replace 'l' -> 'r': " << r << "\n";

    // --- Comparison ---
    std::string a = "apple", b = "apple";
    std::cout << "\n\"apple\" == \"apple\": " << (a == b) << "\n";  // 1 (content)
    std::cout << "compare(\"apple\",\"banana\"): "
              << std::string("apple").compare("banana") << "\n";     // negative

    // --- Iterate ---
    std::cout << "\nChars: ";
    for (char c : s) std::cout << c;
    std::cout << "\n";

    // --- Split ---
    std::string csv = "one,two,three,four";
    std::cout << "\nSplit by ',':\n";
    for (const auto& p : split(csv, ',')) std::cout << "  " << p << "\n";

    // --- Build with += (amortised O(1)) ---
    std::string built;
    for (int i = 1; i <= 5; ++i) {
        built += std::to_string(i);
        if (i < 5) built += "-";
    }
    std::cout << "\nBuilt: " << built << "\n";

    // --- Reverse ---
    std::string rev = "hello";
    std::reverse(rev.begin(), rev.end());
    std::cout << "Reversed 'hello': " << rev << "\n";

    return 0;
}

/*
 * NOTES — C++ vs Java:
 *     - std::string is mutable; += is O(1) amortised — no StringBuilder needed.
 *     - std::string::find returns npos (size_t max) when missing.
 *     - There is no built-in split — use stringstream + getline.
 *     - char operations use <cctype> with unsigned-char casts to avoid UB.
 */

File: Week 7/rust/1.string_basics.rs
/*
 * WEEK 7 - RUST DSA
 * Topic: String Basics
 * File: 1.string_basics.rs
 *
 * CONCEPT:
 *     Rust has two main string types:
 *         &str  — borrowed UTF-8 string slice
 *         String — owned, heap-allocated, growable UTF-8 buffer
 *     Indexing by *byte* offset is allowed only at character boundaries;
 *     for character access prefer .chars().nth(i) (which is O(i)).
 *
 * KEY POINTS:
 *     - Strings are UTF-8; .len() returns BYTES, not chars.
 *     - .chars() yields a char iterator; .bytes() yields u8 bytes.
 *     - .push_str / .push grow a String in O(1) amortised.
 *     - .replace returns a new String; in-place edit needs Vec<char>.
 *
 * ALGORITHM / APPROACH:
 *     Demonstrates: len, chars().nth, find, contains, to_lowercase,
 *     to_uppercase, replace, trim, split, building with String/push_str.
 *
 * RUST-SPECIFIC NOTES:
 *     - .find returns Option<usize>; pattern can be &str or |c| closure.
 *     - For repeated growth use String::with_capacity to preallocate.
 *     - To reverse a UTF-8 string: .chars().rev().collect::<String>().
 *
 * DRY RUN:
 *     s = "Hello, World!"
 *     len = 13 bytes (also 13 chars in this ASCII case)
 *     chars().nth(7) = Some('W')
 *     s[7..12] = "World"     (byte slice; valid here because ASCII)
 *
 * COMPLEXITY:
 *     Most APIs O(n) in input length; chars().nth(i) is O(i).
 */

fn main() {
    let s: &str = "Hello, World!";
    println!("String: {}", s);
    println!("len (bytes): {}", s.len());
    println!("char at 7: {:?}", s.chars().nth(7));      // Some('W')
    println!("byte slice s[7..12]: {}", &s[7..12]);     // "World"
    println!("find('o'): {:?}", s.find('o'));            // Some(4)
    println!("rfind('o'): {:?}", s.rfind('o'));          // Some(8)
    println!("contains(\"World\"): {}", s.contains("World"));
    println!("to_lowercase: {}", s.to_lowercase());
    println!("to_uppercase: {}", s.to_uppercase());
    println!("replace('l',\"r\"): {}", s.replace('l', "r"));
    println!("trim of \"  hello  \": \"{}\"", "  hello  ".trim());

    // --- Comparison ---
    let a: &str = "apple";
    let b: &str = "apple";
    let c: String = String::from("apple");
    println!("\n\"apple\" == \"apple\": {}", a == b);
    println!("\"apple\" == String(\"apple\"): {}", a == c);   // PartialEq compares text
    println!("\"apple\" < \"banana\": {}", "apple" < "banana");

    // --- Iterate chars ---
    print!("\nChars: ");
    for ch in s.chars() {
        print!("{} ", ch);
    }
    println!();

    // --- Split ---
    let csv = "one,two,three,four";
    println!("\nSplit by ',':");
    for token in csv.split(',') {
        println!("  {}", token);
    }

    // --- Build (StringBuilder equivalent) ---
    let mut built = String::with_capacity(16);
    for i in 1..=5 {
        built.push_str(&i.to_string());
        if i < 5 {
            built.push('-');
        }
    }
    println!("\nBuilt: {}", built);

    // --- Reverse ---
    let rev: String = "hello".chars().rev().collect();
    println!("Reversed 'hello': {}", rev);
}

/*
 * NOTES — Rust vs Java:
 *     - String/&str distinction enforces ownership: prefer &str for params.
 *     - .len() is bytes — for char count use .chars().count().
 *     - String is mutable and amortised O(1) for push/push_str.
 *     - PartialOrd on &str is lexicographic; PartialEq compares content.
 */
